src/pil/backend/powdr_pil.rs

```diff
@@ -0,0 +1,488 @@
+        writeln!(pil, "// === Witness Columns for Signals ===").unwrap();
+        let mut col_witness = String::from("col witness ");
+
+        let mut col_witness_vars = pil_ir
+            .col_witness
+            .iter()
+            .map(|uuid| annotations_map.get(uuid).unwrap().clone())
+            .collect::<Vec<String>>();
+
+        // Get unique witness column annotations, because the same internal signal across
+        // different step types have different UUIDs and therefore appear multiple times.
+        col_witness_vars.sort();
+        col_witness_vars.dedup();
+        col_witness = col_witness + col_witness_vars.join(", ").as_str() + ";";
+        writeln!(pil, "{}", col_witness).unwrap();
+    }
+}
+
+fn generate_pil_fixed_columns<F: Debug>(
+    pil: &mut String,
+    pil_ir: &PilCircuit<F>,
+    annotations_map: &HashMap<UUID, String>,
+) {
+    if !pil_ir.fixed_signals.is_empty() {
+        writeln!(pil, "// === Fixed Columns for Signals ===").unwrap();
+        for (fixed_signal_id, assignments) in pil_ir.fixed_signals.iter() {
+            let fixed_name = annotations_map.get(fixed_signal_id).unwrap().clone();
+            let mut assignments_string = String::new();
+            let assignments_vec = assignments
+                .iter()
+                .map(|assignment| format!("{:?}", assignment))
+                .collect::<Vec<String>>();
+            write!(
+                assignments_string,
+                "{}",
+                assignments_vec.join(", ").as_str()
+            )
+            .unwrap();
+            writeln!(pil, "col fixed {} = [{}];", fixed_name, assignments_string).unwrap();
+        }
+    }
+}
+
+fn generate_pil_step_selector_columns<F: Clone>(
+    pil: &mut String,
+    pil_ir: &PilCircuit<F>,
+    annotations_map: &HashMap<UUID, String>,
+) {
+    if !pil_ir.step_types_instantiations.is_empty() {
+        writeln!(pil, "// === Fixed Columns for Step Type Selectors ===").unwrap();
+        for (step_type_id, step_type_instantiation) in pil_ir.step_types_instantiations.iter() {
+            let step_type_name = annotations_map.get(step_type_id).unwrap().clone();
+            let mut step_type_selector = String::new();
+            write!(step_type_selector, "col fixed {} = [", step_type_name).unwrap();
+
+            for (index, step_instance) in step_type_instantiation.iter().enumerate() {
+                write!(step_type_selector, "{}", step_instance).unwrap();
+                if index == pil_ir.num_steps - 1 {
+                    write!(step_type_selector, "]").unwrap();
+                } else {
+                    write!(step_type_selector, ", ").unwrap();
+                }
+            }
+            writeln!(pil, "{}", step_type_selector).unwrap();
+        }
+    }
+}
+
+fn generate_pil_step_types<F: Debug + Clone>(
+    pil: &mut String,
+    pil_ir: &PilCircuit<F>,
+    annotations_map: &HashMap<UUID, String>,
+) {
+    if !pil_ir.step_types.is_empty() {
```

Comment:
If the step selector columns for PIL are normal advice or fixed columns, then they should be in the IR as normal PIL columns

Reply:
Agreed and fixed.